core/src/main/scala/org/apache/spark/sql/delta/commands/RestoreTableCommand.scala

```diff
@@ -0,0 +1,184 @@
+ *  6) Commit metadata, Protocol, all RemoveFile and AddFile actions
+ *  into delta log using `commitLarge`.
+ *  7) If table was modified in parallel then ignore restore and raise exception.
+ */
+case class RestoreTableCommand(
+  deltaLog: DeltaLog,
+  version: Option[Long],
+  timestamp: Option[Timestamp]
+) extends LeafRunnableCommand with DeltaCommand {
+
+  override def run(spark: SparkSession): Seq[Row] = {
+    recordDeltaOperation(deltaLog, "delta.restore") {
+
+      require(version.isEmpty ^ timestamp.isEmpty,
+        "Either the version or timestamp should be provided for restore")
+      val parallelism = restoreParallelism(spark)
+      val latestSnapshot = deltaLog.update()
+      val versionToRestore = version.getOrElse(
+        deltaLog
+          .history
+          .getActiveCommitAtTime(timestamp.get, canReturnLastCommit = true)
+          .version
+      )
+
+      require(versionToRestore < latestSnapshot.version,
+        s"Version to restore ($versionToRestore) should be less then " +
+        s"last available version (${latestSnapshot.version})")
+
+      val snapshotToRestore = deltaLog.getSnapshotAt(versionToRestore)
+      val latestSnapshotFiles = latestSnapshot.allFiles
+      val snapshotToRestoreFiles = snapshotToRestore.allFiles
+
+      import spark.implicits._
+      import collection.JavaConverters._
+
+      val filesToAdd = snapshotToRestoreFiles
+        .join(
+          latestSnapshotFiles,
+          snapshotToRestoreFiles("path") ===  latestSnapshotFiles("path"),
+          "left_anti")
+        .as[AddFile]
+        .map(_.copy(dataChange = true))
+        .repartition(parallelism)
+        .cache() // To avoid Dataset recompute for each partition of toLocalIterator()
+
+      checkSnapshotFilesAvailability(deltaLog, filesToAdd, versionToRestore)
+
+      val filesToRemove = latestSnapshotFiles
+        .join(
+          snapshotToRestoreFiles,
+          latestSnapshotFiles("path") ===  snapshotToRestoreFiles("path"),
+          "left_anti")
+        .as[AddFile]
+        .map(_.removeWithTimestamp())
+        .repartition(parallelism)
+        .cache() // To avoid Dataset recompute for each partition of toLocalIterator()
+
+      // Commit files, metrics, protocol and metadata to delta log
+      deltaLog.withNewTransaction { txn =>
+
+        val metrics = computeMetrics(filesToAdd, filesToRemove, snapshotToRestore)
+        val addActions = filesToAdd.toLocalIterator().asScala
+        val removeActions = filesToRemove.toLocalIterator().asScala
+
+        txn.updateMetadata(snapshotToRestore.metadata)
+
+        commitLarge(
+          spark,
+          txn,
+          addActions ++ removeActions,
+          DeltaOperations.Restore(version, timestamp.map(_.getTime)),
+          Map.empty,
+          metrics)
+      }
+      filesToAdd.unpersist()
+      filesToRemove.unpersist()
+
+      Seq.empty[Row]
+    }
+  }
+
+  private def restoreParallelism(spark: SparkSession): Int = spark
```

Comment:
why do you need a separate parameter for the parallelism. why no using the spark.sql.shuffle.partitions... which is the standard way every one configures parallelism in spark shuffles

Reply:
The idea was to be more flexible with this parameter (for example, avoid affecting of whole spark share cluster). But I agree that we can use spark.sql.shuffle.partitions :+1:

Reply:
yeah. its a delicate tradeoff. having special-purpose conf does allow fine tuning, but only if you know. and conf discovery is a major bottleneck for users. so its usually not useful if there is a conf but no one knows about it. rather it is more annoying to the user if the related confs that they know about do not work because they are not being used. so we always prefer to minimize surprises / annoyance by reusing confs as much as possible. :) thank you for understanding.